github_slack_bot/utils/slack.py

```diff
@@ -0,0 +1,145 @@
+                "ok": False,
+                "error": "Failed to parse message. Issue title and issue body should be "
+                         "separated by a double underscore (__)"
+            }
+
+        logger.info(res)
+        return res
+
+    def __init__(self, slack_api_token=None, git_client=None):
+        self.git_client = git_client
+        self.slack_api_token = slack_api_token
+        self.bot_user_id = None
+        self.sc = None
+
+        logger.info("Slack bot initialized")
+
+    def set_bot_user_id(self):
+        """Make an auth.test Slack API call to retrieve the bot's user_id.
+        Stores the user_id in the "bot_user_id" instance attribute
+        :return: None
+        """
+        auth_test_res = self.sc.api_call("auth.test")
+        if auth_test_res["ok"]:
+            self.bot_user_id = auth_test_res['user_id']
+            logger.info("Found bot's user_id: {}".format(self.bot_user_id))
+        else:
+            auth_test_error = auth_test_res["error"]
+            logger.error("Failed to retrieve bot's user_id due to: {}"
+                         .format(auth_test_error))
+            raise Exception(EXCEPTION_DEFAULT_MSG)
+
+    # TODO: extend handle_rtm_event() to handle different commands, such as "update-issue", "open-pr"
+    def handle_rtm_event(self, event):
+        """Processes and handles RTM events
+        :param event: Slack RTM event
+        :return: None
+        """
+        if "type" in event and event["type"] == "message" and "subtype" not in event:
+            msg = event['text']
+            channel = event["channel"]
+
+            if SlackBot.is_bot_message(self.bot_user_id, msg):
+                logger.info("Received a message "
+                            "mentioning our bot: '{}'".format(msg))
+                parsed_msg = SlackBot.parse_bot_message(msg)
+
+                if parsed_msg["ok"]:
+                    git_res = self.git_client.create_issue(parsed_msg["issue_title"],
+                                                           parsed_msg["issue_body"])
+                    if git_res["ok"]:
+                        res = "GitHub issue created successfully: {}".format(git_res["link"])
+                    else:
+                        res = "Failed to create issue: {}".format(git_res["error"])
+                else:
+                    res = parsed_msg["error"]
+
+                self.sc.rtm_send_message(channel, res)
+
+    # TODO: change bot implementation to use events API so we can scale horizontally
+    def start_bot(self):
+        """Creates a connection to the RTM API and listens for events in an infinite loop
+        :return: None
+        """
+        self.sc = SlackClient(self.slack_api_token)
+        self.set_bot_user_id()
+
+        if self.sc.rtm_connect(auto_reconnect=True):
+            logger.info("Slack client connected to RTM API")
+            while True:
+                try:
+                    slack_events = self.sc.rtm_read()
```

Comment:
Does this block until a read is available? Or how does this work?

Reply:
It doesn't block, it reads from the RTM stream and returns and empty list if there are no events. I added a delay between reads as shown in their docs here: https://slack.dev/python-slackclient/real_time_messaging.html#connecting-to-the-rtm-api